reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,487 @@
+                    .flatMap(negated -> iterate(negated.disjunction().conjunctions()))
+                    .map(negatedConjunction -> {
+                        answerCountEstimator.registerConjunctionAndInitialize(negatedConjunction);
+                        return answerCountEstimator.estimateAllAnswers(negatedConjunction);
+                    }).reduce(0L, Long::sum);
+        }
+
+        private static abstract class LocalEstimate {
+
+            final List<Variable> variables;
+
+            private LocalEstimate(List<Variable> variables) {
+                this.variables = variables;
+            }
+
+            abstract long answerEstimate(Set<Variable> variableFilter);
+
+            private static class SimpleEstimate extends LocalEstimate {
+                private final long staticEstimate;
+
+                private SimpleEstimate(List<Variable> variables, long estimate) {
+                    super(variables);
+                    this.staticEstimate = estimate;
+                }
+
+                @Override
+                long answerEstimate(Set<Variable> variableFilter) {
+                    return staticEstimate;
+                }
+            }
+
+            public static class CoPlayerEstimate extends LocalEstimate {
+
+                private final Map<TypeVariable, Integer> rolePlayerCounts;
+                private final Map<TypeVariable, Long> rolePlayerEstimates;
+                private final double relationTypeEstimate;
+                private final long inferredRelationEstimate;
+
+                public CoPlayerEstimate(List<Variable> variables, double relationTypeEstimate,
+                                        Map<TypeVariable, Long> rolePlayerEstimates, Map<TypeVariable, Integer> rolePlayerCounts,
+                                        long inferredRelationEstimate) {
+                    super(variables);
+                    this.relationTypeEstimate = relationTypeEstimate;
+                    this.rolePlayerEstimates = rolePlayerEstimates;
+                    this.rolePlayerCounts = rolePlayerCounts;
+                    this.inferredRelationEstimate = inferredRelationEstimate;
+                }
+
+                @Override
+                long answerEstimate(Set<Variable> variableFilter) {
+                    long singleRelationEstimate = 1L;
+                    for (TypeVariable key : rolePlayerCounts.keySet()) {
+                        assert rolePlayerEstimates.containsKey(key);
+                        long avgRolePlayers = Double.valueOf(Math.ceil(rolePlayerEstimates.get(key) / relationTypeEstimate)).longValue();
+                        singleRelationEstimate *= nPermuteKforSmallK(avgRolePlayers, rolePlayerCounts.get(key));
+                    }
+
+                    long fullEstimate = Double.valueOf(Math.ceil(relationTypeEstimate * singleRelationEstimate)).longValue() + inferredRelationEstimate;
+                    return fullEstimate;
+                }
```

Comment:
Store just enough information to be able to query answers to subset of variables.